In [1]:
#Import the libraries
from sklearn.cluster import DBSCAN
from sklearn.preprocessing import StandardScaler

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Import .xlsx file and convert it to a DataFrame object
from google.colab import files
uploaded = files.upload()

Saving EastWestAirlines.xlsx to EastWestAirlines.xlsx


In [4]:
data=pd.read_excel('EastWestAirlines.xlsx')

In [5]:
data

ID#  Balance  Qual_miles  cc1_miles  cc2_miles  cc3_miles  Bonus_miles  \
0        1    28143           0          1          1          1          174   
1        2    19244           0          1          1          1          215   
2        3    41354           0          1          1          1         4123   
3        4    14776           0          1          1          1          500   
4        5    97752           0          4          1          1        43300   
...    ...      ...         ...        ...        ...        ...          ...   
3994  4017    18476           0          1          1          1         8525   
3995  4018    64385           0          1          1          1          981   
3996  4019    73597           0          3          1          1        25447   
3997  4020    54899           0          1          1          1          500   
3998  4021     3016           0          1          1          1            0   

      Bonus_trans  Flight_miles_12mo  Flight_trans_12  Days_since_enroll  \
0               1                  0                0               7000   
1               2                  0                0               6968   
2               4                  0                0               7034   
3               1                  0                0               6952   
4              26               2077                4               6935   
...           ...                ...              ...                ...   
3994            4                200                1               1403   
3995            5                  0                0               1395   
3996            8                  0                0               1402   
3997            1                500                1               1401   
3998            0                  0                0               1398   

      Award?  
0          0  
1          0  
2          0  
3          0  
4          1  
...      ...  
3994       1  
3995       1  
3996       1  
3997       0  
3998       0  

[3999 rows x 12 columns]

In [6]:
data.shape

(3999, 12)

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   ID#                3999 non-null   int64
 1   Balance            3999 non-null   int64
 2   Qual_miles         3999 non-null   int64
 3   cc1_miles          3999 non-null   int64
 4   cc2_miles          3999 non-null   int64
 5   cc3_miles          3999 non-null   int64
 6   Bonus_miles        3999 non-null   int64
 7   Bonus_trans        3999 non-null   int64
 8   Flight_miles_12mo  3999 non-null   int64
 9   Flight_trans_12    3999 non-null   int64
 10  Days_since_enroll  3999 non-null   int64
 11  Award?             3999 non-null   int64
dtypes: int64(12)
memory usage: 375.0 KB


In [8]:
data.isnull().sum()

ID#                  0
Balance              0
Qual_miles           0
cc1_miles            0
cc2_miles            0
cc3_miles            0
Bonus_miles          0
Bonus_trans          0
Flight_miles_12mo    0
Flight_trans_12      0
Days_since_enroll    0
Award?               0
dtype: int64

In [9]:
data.columns

Index(['ID#', 'Balance', 'Qual_miles', 'cc1_miles', 'cc2_miles', 'cc3_miles',
       'Bonus_miles', 'Bonus_trans', 'Flight_miles_12mo', 'Flight_trans_12',
       'Days_since_enroll', 'Award?'],
      dtype='object')

#Dropping unnecessary feaures

In [10]:
data1=data.drop(["ID#"],axis=1)

In [11]:
data1

Balance  Qual_miles  cc1_miles  cc2_miles  cc3_miles  Bonus_miles  \
0       28143           0          1          1          1          174   
1       19244           0          1          1          1          215   
2       41354           0          1          1          1         4123   
3       14776           0          1          1          1          500   
4       97752           0          4          1          1        43300   
...       ...         ...        ...        ...        ...          ...   
3994    18476           0          1          1          1         8525   
3995    64385           0          1          1          1          981   
3996    73597           0          3          1          1        25447   
3997    54899           0          1          1          1          500   
3998     3016           0          1          1          1            0   

      Bonus_trans  Flight_miles_12mo  Flight_trans_12  Days_since_enroll  \
0               1                  0                0               7000   
1               2                  0                0               6968   
2               4                  0                0               7034   
3               1                  0                0               6952   
4              26               2077                4               6935   
...           ...                ...              ...                ...   
3994            4                200                1               1403   
3995            5                  0                0               1395   
3996            8                  0                0               1402   
3997            1                500                1               1401   
3998            0                  0                0               1398   

      Award?  
0          0  
1          0  
2          0  
3          0  
4          1  
...      ...  
3994       1  
3995       1  
3996       1  
3997       0  
3998       0  

[3999 rows x 11 columns]

#Renaming the features

In [14]:
data2=data1.rename(columns={'Qual_miles':'Qual','cc1_miles':'cc1','cc2_miles':'cc2','cc3_miles':'cc1','Flight_trans_12':'FT','Days_since_enroll':'Days','Award?':'Award','Bonus_miles':'BM','Bonus_trans':'BT','Flight_miles_12mo':'FM'})
data2

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award
0       28143     0    1    1    1    174   1     0   0  7000      0
1       19244     0    1    1    1    215   2     0   0  6968      0
2       41354     0    1    1    1   4123   4     0   0  7034      0
3       14776     0    1    1    1    500   1     0   0  6952      0
4       97752     0    4    1    1  43300  26  2077   4  6935      1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...
3994    18476     0    1    1    1   8525   4   200   1  1403      1
3995    64385     0    1    1    1    981   5     0   0  1395      1
3996    73597     0    3    1    1  25447   8     0   0  1402      1
3997    54899     0    1    1    1    500   1   500   1  1401      0
3998     3016     0    1    1    1      0   0     0   0  1398      0

[3999 rows x 11 columns]

In [15]:
array=data2.values
array

array([[28143,     0,     1, ...,     0,  7000,     0],
       [19244,     0,     1, ...,     0,  6968,     0],
       [41354,     0,     1, ...,     0,  7034,     0],
       ...,
       [73597,     0,     3, ...,     0,  1402,     1],
       [54899,     0,     1, ...,     1,  1401,     0],
       [ 3016,     0,     1, ...,     0,  1398,     0]])

In [16]:
array

array([[28143,     0,     1, ...,     0,  7000,     0],
       [19244,     0,     1, ...,     0,  6968,     0],
       [41354,     0,     1, ...,     0,  7034,     0],
       ...,
       [73597,     0,     3, ...,     0,  1402,     1],
       [54899,     0,     1, ...,     1,  1401,     0],
       [ 3016,     0,     1, ...,     0,  1398,     0]])

In [17]:
stscaler = StandardScaler()
X = stscaler.fit_transform(array)

In [18]:
X

array([[-4.51140783e-01, -1.86298687e-01, -7.69578406e-01, ...,
        -3.62167870e-01,  1.39545434e+00, -7.66919299e-01],
       [-5.39456874e-01, -1.86298687e-01, -7.69578406e-01, ...,
        -3.62167870e-01,  1.37995704e+00, -7.66919299e-01],
       [-3.20031232e-01, -1.86298687e-01, -7.69578406e-01, ...,
        -3.62167870e-01,  1.41192021e+00, -7.66919299e-01],
       ...,
       [-4.29480975e-05, -1.86298687e-01,  6.83121167e-01, ...,
        -3.62167870e-01, -1.31560393e+00,  1.30391816e+00],
       [-1.85606976e-01, -1.86298687e-01, -7.69578406e-01, ...,
        -9.85033311e-02, -1.31608822e+00, -7.66919299e-01],
       [-7.00507951e-01, -1.86298687e-01, -7.69578406e-01, ...,
        -3.62167870e-01, -1.31754109e+00, -7.66919299e-01]])

In [24]:
dbscan = DBSCAN(eps=0.5, min_samples=3)
dbscan.fit_predict(X)

array([0, 0, 0, ..., 6, 0, 0])

In [25]:
#Noisy samples are given the label -1.
dbscan.labels_

array([0, 0, 0, ..., 6, 0, 0])

In [26]:
cl=pd.DataFrame(dbscan.labels_,columns=['cluster'])
cl

cluster
0           0
1           0
2           0
3           0
4          -1
...       ...
3994        1
3995        1
3996        6
3997        0
3998        0

[3999 rows x 1 columns]

In [28]:
clustered = pd.concat([data2,cl],axis=1)

In [29]:
clustered

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
0       28143     0    1    1    1    174   1     0   0  7000      0        0
1       19244     0    1    1    1    215   2     0   0  6968      0        0
2       41354     0    1    1    1   4123   4     0   0  7034      0        0
3       14776     0    1    1    1    500   1     0   0  6952      0        0
4       97752     0    4    1    1  43300  26  2077   4  6935      1       -1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3994    18476     0    1    1    1   8525   4   200   1  1403      1        1
3995    64385     0    1    1    1    981   5     0   0  1395      1        1
3996    73597     0    3    1    1  25447   8     0   0  1402      1        6
3997    54899     0    1    1    1    500   1   500   1  1401      0        0
3998     3016     0    1    1    1      0   0     0   0  1398      0        0

[3999 rows x 12 columns]

In [30]:
clustered[clustered['cluster']==-1]

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
4       97752     0    4    1    1  43300  26  2077   4  6935      1       -1
6       84914     0    3    1    1  27482  25     0   0  6994      0       -1
8      443003     0    3    2    1   1753  43  3850  12  6948      1       -1
9      104860     0    3    1    1  28426  28  1150   3  6931      1       -1
19      23354     0    3    1    1  10447   5     0   0  6896      1       -1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3961   131125   862    1    1    1   4993  11    50   1  1443      0       -1
3967    26504     0    2    1    1  10599  29  1000   7  1445      0       -1
3978    10071     0    2    1    1  27701  16     0   0  1421      1       -1
3979    57793     0    3    1    1  20959  15  1198   3  1424      1       -1
3985    59017     0    4    1    1  34746  25     0   0  1418      0       -1

[1228 rows x 12 columns]

so there are 1228 rows contains outliers

In [31]:
clustered

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
0       28143     0    1    1    1    174   1     0   0  7000      0        0
1       19244     0    1    1    1    215   2     0   0  6968      0        0
2       41354     0    1    1    1   4123   4     0   0  7034      0        0
3       14776     0    1    1    1    500   1     0   0  6952      0        0
4       97752     0    4    1    1  43300  26  2077   4  6935      1       -1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3994    18476     0    1    1    1   8525   4   200   1  1403      1        1
3995    64385     0    1    1    1    981   5     0   0  1395      1        1
3996    73597     0    3    1    1  25447   8     0   0  1402      1        6
3997    54899     0    1    1    1    500   1   500   1  1401      0        0
3998     3016     0    1    1    1      0   0     0   0  1398      0        0

[3999 rows x 12 columns]

In [32]:
clustered.mean()

Balance    73601.327582
Qual         144.114529
cc1            2.059515
cc2            1.014504
cc1            1.012253
BM         17144.846212
BT            11.601900
FM           460.055764
FT             1.373593
Days        4118.559390
Award          0.370343
cluster        1.997999
dtype: float64

#Few more iterations for different values of epsilon and sample

#performing an iteration for epsilon value 1 and min sample 4

In [33]:
dbscan = DBSCAN(eps=1, min_samples=4)
dbscan.fit_predict(X)

array([0, 0, 0, ..., 1, 0, 0])

In [34]:
#Noisy samples are given the label -1.
dbscan.labels_

array([0, 0, 0, ..., 1, 0, 0])

In [35]:
cl=pd.DataFrame(dbscan.labels_,columns=['cluster'])
cl

cluster
0           0
1           0
2           0
3           0
4           1
...       ...
3994        1
3995        1
3996        1
3997        0
3998        0

[3999 rows x 1 columns]

In [36]:
clustered = pd.concat([data2,cl],axis=1)

In [37]:
clustered

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
0       28143     0    1    1    1    174   1     0   0  7000      0        0
1       19244     0    1    1    1    215   2     0   0  6968      0        0
2       41354     0    1    1    1   4123   4     0   0  7034      0        0
3       14776     0    1    1    1    500   1     0   0  6952      0        0
4       97752     0    4    1    1  43300  26  2077   4  6935      1        1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3994    18476     0    1    1    1   8525   4   200   1  1403      1        1
3995    64385     0    1    1    1    981   5     0   0  1395      1        1
3996    73597     0    3    1    1  25447   8     0   0  1402      1        1
3997    54899     0    1    1    1    500   1   500   1  1401      0        0
3998     3016     0    1    1    1      0   0     0   0  1398      0        0

[3999 rows x 12 columns]

In [38]:
clustered[clustered['cluster']==-1]

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
8      443003     0    3    2    1   1753  43  3850  12  6948      1       -1
21     185681  2024    1    1    1  13300  16  1800   9  6896      1       -1
22      20584     0    1    1    1   3450  11  3450  11  6884      0       -1
24     205651   500    1    1    1   4025  21   700   4  7932      0       -1
32     276571     0    4    1    1  42044  23     0   0  7872      0       -1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3911    55922  2273    1    1    1   8104  19  4104  16  1552      0       -1
3917   130748     0    4    1    1  65510  37  6349   7  1545      0       -1
3934    76595     0    1    1    1  10577   2  9577   1  1515      0       -1
3961   131125   862    1    1    1   4993  11    50   1  1443      0       -1
3967    26504     0    2    1    1  10599  29  1000   7  1445      0       -1

[529 rows x 12 columns]

#Interpretation:we can observe here than by taking epsilon value 1 as and min sample values 4 we getting outliers in 529 rows . which means as we increase epsilon value i.e radius number of outliers decreases.

In [40]:
clustered.mean()

Balance    73601.327582
Qual         144.114529
cc1            2.059515
cc2            1.014504
cc1            1.012253
BM         17144.846212
BT            11.601900
FM           460.055764
FT             1.373593
Days        4118.559390
Award          0.370343
cluster        0.186297
dtype: float64

#Here iam perfroming an iteration for epsilon radius of 2 and sample size 4

In [41]:
dbscan = DBSCAN(eps=2, min_samples=4)
dbscan.fit_predict(X)

array([0, 0, 0, ..., 1, 0, 0])

In [42]:
#Noisy samples are given the label -1.
dbscan.labels_

array([0, 0, 0, ..., 1, 0, 0])

In [43]:
cl=pd.DataFrame(dbscan.labels_,columns=['cluster'])
cl

cluster
0           0
1           0
2           0
3           0
4           1
...       ...
3994        1
3995        1
3996        1
3997        0
3998        0

[3999 rows x 1 columns]

In [44]:
clustered = pd.concat([data2,cl],axis=1)

In [45]:
clustered

Balance  Qual  cc1  cc2  cc1     BM  BT    FM  FT  Days  Award  cluster
0       28143     0    1    1    1    174   1     0   0  7000      0        0
1       19244     0    1    1    1    215   2     0   0  6968      0        0
2       41354     0    1    1    1   4123   4     0   0  7034      0        0
3       14776     0    1    1    1    500   1     0   0  6952      0        0
4       97752     0    4    1    1  43300  26  2077   4  6935      1        1
...       ...   ...  ...  ...  ...    ...  ..   ...  ..   ...    ...      ...
3994    18476     0    1    1    1   8525   4   200   1  1403      1        1
3995    64385     0    1    1    1    981   5     0   0  1395      1        1
3996    73597     0    3    1    1  25447   8     0   0  1402      1        1
3997    54899     0    1    1    1    500   1   500   1  1401      0        0
3998     3016     0    1    1    1      0   0     0   0  1398      0        0

[3999 rows x 12 columns]

In [46]:
clustered[clustered['cluster']==-1]

Balance  Qual  cc1  cc2  cc1      BM  BT    FM  FT  Days  Award  cluster
8      443003     0    3    2    1    1753  43  3850  12  6948      1       -1
50     108137     0    1    1    1    6368   5  6368   5  6844      0       -1
65      53914     0    3    1    1   33767  45  5550  29  6826      1       -1
70      84409  5031    2    1    1   15436  16  1150   4  7766      0       -1
87     609477     0    3    1    1   21422  22  1200   8  6820      0       -1
...       ...   ...  ...  ...  ...     ...  ..   ...  ..   ...    ...      ...
3846    97510  1678    5    1    1   71609  51  7650  26  1665      1       -1
3883   126167     0    5    1    1  190160  15     0   0  1612      1       -1
3911    55922  2273    1    1    1    8104  19  4104  16  1552      0       -1
3917   130748     0    4    1    1   65510  37  6349   7  1545      0       -1
3934    76595     0    1    1    1   10577   2  9577   1  1515      0       -1

[167 rows x 12 columns]

As we can here by taking epsilon value of 2 and min sample size and 2  we are getting outlier 167 rows.